ICS3 connection versioning / model checking in TLA+ (#211)
informalsystems/ibc-rs: docs/disclosure-log.md

Chain `B` drops this message.

From this point on, the model stutters, i.e., is unable to progress further in the connection handshake protocol.

### 3. ICS3 problems due to version negotiation

__Context__.
The original issue triggering this discussion is here: [cosmos/ics/#459](https://github.com/cosmos/ics/issues/459).

> Briefly, version negotiation in the ICS3 handshake can interfere in various ways, breaking either the safety or liveness of this protocol.

Reviewer: The current liveness issues are expected as the two parties do not agree on a version. Not sure if and how these should be reported. If there would be liveness issue due to protocol bugs (e.g. the two parties do not agree even if there is one common version) the format here would be fine.

Several solution candidates exist, which we classify by their "mode", i.e., a strategy for picking the version at some point or another in the protocol.
For a full description of the modes, please consult [L2-tla/readme.md#version-negotiation-modes](L2-tla/README.md#version-negotiation-modes).

__Overview__.
Below we use TLA+ traces to explore and report on the exact problems that can occur. We also show how the solution candidates fare.
These are the takeaways from the discussion below:

1. The set of compatible versions that chains start off with (return values of `getCompatibleVersions()` in ICS3) have to intersect, otherwise a liveness issue occurs. This assumption is independent of the version negotiation mode. We report this in __case (a)__ below.
2. Modes "overwrite", "onTryNonDet", and "onAckNonDet" all result in breaking the handshake protocol. See __cases (b), (c), and (d)__ below for traces.
3. The deterministic modes "onTryDet" and "onAckDet" pass model checking, so a solution should be chosen among these two candidates (see the original issue for follow-up on the solution).

> ##### Case (a). Empty version intersection causes liveness issue

Reviewer: Could you add a summary table with the different modes and check which have problems and what kind? I think it would be useful to give an overall view of the per mode issues and also a place where we see differences among modes.

This message has proofs for height `2`, so the environment also does a `UpdateClient` on chain `A` for consensus state at height `2`.
With this update, chain `A` advances to height `5`.

9. Chain `A` processes the `ICS3MsgTry`, advances to height `6`, and prepares a `ICS3MsgAck` for chain `B`.
The version in this message is `<<2>>`, which is the version which chain `A` choose non-deterministically for this connection.
The connection on chain `A` is now in state `TRYOPEN`.

10. Chain `B` processes the `ICS3MsgTry`, advances to height `5`, and prepares a `ICS3MsgAck` for chain `A`.
The version in this message is `<<1>>`, which chain `B` choose non-deterministically for this connection.
The connection on chain `B` is now in state `TRYOPEN`.

From this point on, the two chains can not make further progress in the handshake, since they chose different versions.
Neither of the two chains can process the `ICS3MsgAck` message because the version in this message does not match with the version the chain stores locally.
(A chain should not overwrite its local version either, otherwise the safety issue from case (b) can appear.)
Therefore, the model stutters (cannot progress anymore).

> ##### Case (d). Mode `onAckNonDet` causes liveness issue

Reviewer: ?
```suggestion
##### Case (d). Mode `onAckNonDet` causes safety issue
```

Model checking details in TLA+:
- Model parameters:
```
Concurrency <- FALSE
MaxBufLen <- 2
MaxHeight <- 7
MaxVersionNr <- 2
VersionPickMode <- "onAckNonDet"
```
- Check for invariant _VersionInvariant, as well as _Deadlock_ and property _Termination_.

Outcome:

> - Model checking halts with exception "Temporal properties were violated."

Reviewer: I expect this:
```suggestion
- Model checking halts with exception "Invariant VersionInvariant is violated."
```
